File: v26_mediator/java/GUIMain.java
package v26_mediator.java;

// Mediator interface
interface Mediator {
    void notify(Component sender, String event);
}

// Concrete Mediator
class ConcreteMediator implements Mediator {
    private Button button;
    private TextBox textBox;
    private Label label;

    public ConcreteMediator(Button button, TextBox textBox, Label label) {
        this.button = button;
        this.button.setMediator(this);
        this.textBox = textBox;
        this.textBox.setMediator(this);
        this.label = label;
        this.label.setMediator(this);
    }

    @Override
    public void notify(Component sender, String event) {
        if (sender == button && event.equals("clicked")) {
            textBox.clear();
            label.show("Button was clicked");
        }
    }
}

// Abstract Colleague
abstract class Component {
    protected Mediator mediator;

    public void setMediator(Mediator mediator) {
        this.mediator = mediator;
    }

    public void changed(String event) {
        mediator.notify(this, event);
    }
}

// Button Colleague
class Button extends Component {
    public void click() {
        System.out.println("Button clicked");
        changed("clicked");
    }
}

// TextBox Colleague
class TextBox extends Component {
    public void clear() {
        System.out.println("TextBox cleared");
    }
}

// Label Colleague
class Label extends Component {
    public void show(String message) {
        System.out.println("Label shows: " + message);
    }
}

// Main
public class GUIMain {
    public static void main(String[] args) {
        Button button = new Button();
        TextBox textBox = new TextBox();
        Label label = new Label();
        new ConcreteMediator(button, textBox, label);

        // Simulate a button click
        button.click();
    }
}

File: v26_mediator/python/gui_main.py
from abc import ABC, abstractmethod

class Mediator(ABC):
    @abstractmethod
    def notify(self, sender, event):
        pass

class ConcreteMediator(Mediator):
    def __init__(self, button, text_box, label):
        self.button = button
        self.button.set_mediator(self)
        self.text_box = text_box
        self.text_box.set_mediator(self)
        self.label = label
        self.label.set_mediator(self)

    def notify(self, sender, event):
        if sender == self.button and event == 'clicked':
            self.text_box.clear()
            self.label.show('Button was clicked')

class Component(ABC):
    def __init__(self):
        self.mediator = None

    def set_mediator(self, mediator):
        self.mediator = mediator

    def changed(self, event):
        self.mediator.notify(self, event)

class Button(Component):
    def click(self):
        print('Button clicked')
        self.changed('clicked')

class TextBox(Component):
    def clear(self):
        print('TextBox cleared')

class Label(Component):
    def show(self, message):
        print(f'Label shows: {message}')

if __name__ == '__main__':
    button = Button()
    text_box = TextBox()
    label = Label()
    ConcreteMediator(button, text_box, label)

    # Simulate a button click
    button.click()


